src/internal/packager/images/pull.go

```diff
@@ -27,329 +31,313 @@ import (
-		referenceToDigest[finishedImage.refInfo.Reference] = finishedImage.digest
-	}
-
-	onImageSavingError := func(err error) error {
-		// Send a signal to the progress bar that we're done and wait for the thread to finish
-		doneSaving <- err
-		<-doneSaving
-		message.WarnErr(err, "Failed to write image config or manifest, trying again up to 3 times...")
+	sc := func() error {
+		saved, err := SaveConcurrent(ctx, cranePath, toPull)
+		for k := range saved {
+			delete(toPull, k)
+		}
 		return err
 	}
 
-	if err := imageSavingConcurrency.WaitWithProgress(onImageSavingProgress, onImageSavingError); err != nil {
-		return nil, err
-	}
-
-	// for every image sequentially append OCI descriptor
-	for refInfo, img := range refInfoToImage {
-		desc, err := partial.Descriptor(img)
-		if err != nil {
-			return nil, err
+	ss := func() error {
+		saved, err := SaveSequential(ctx, cranePath, toPull)
+		for k := range saved {
+			delete(toPull, k)
 		}
+		return err
+	}
 
-		if err := cranePath.AppendDescriptor(*desc); err != nil {
-			return nil, err
-		}
+	if err := helpers.Retry(sc, 2, 5*time.Second, message.Warnf); err != nil {
+		message.Warnf("Failed to save images in parallel, falling back to sequential save: %s", err.Error())
 
-		imgDigest, err := img.Digest()
-		if err != nil {
+		if err := helpers.Retry(ss, 2, 5*time.Second, message.Warnf); err != nil {
 			return nil, err
 		}
-
-		referenceToDigest[refInfo.Reference] = imgDigest.String()
-	}
-
-	if err := utils.AddImageNameAnnotation(i.ImagesPath, referenceToDigest); err != nil {
-		return nil, fmt.Errorf("unable to format OCI layout: %w", err)
 	}
 
 	// Send a signal to the progress bar that we're done and wait for the thread to finish
 	doneSaving <- nil
 	<-doneSaving
 
-	return imgInfoList, nil
+	return fetched, nil
 }
 
-// PullImage returns a v1.Image either by loading a local tarball or pulling from the wider internet.
-func (i *ImageConfig) PullImage(src string, spinner *message.Spinner) (img v1.Image, hasImageLayers bool, err error) {
-	cacheImage := false
-	// Load image tarballs from the local filesystem.
-	if strings.HasSuffix(src, ".tar") || strings.HasSuffix(src, ".tar.gz") || strings.HasSuffix(src, ".tgz") {
-		spinner.Updatef("Reading image tarball: %s", src)
-		img, err = crane.Load(src, config.GetCraneOptions(true, i.Architectures...)...)
-		if err != nil {
-			return nil, false, err
+// CleanupInProgressLayers removes incomplete layers from the cache.
+func CleanupInProgressLayers(ctx context.Context, img v1.Image) error {
+	layers, err := img.Layers()
+	if err != nil {
+		return err
+	}
+	eg, _ := errgroup.WithContext(ctx)
+	for _, layer := range layers {
+		layer := layer
+		eg.Go(func() error {
+			digest, err := layer.Digest()
+			if err != nil {
+				return err
+			}
+			size, err := layer.Size()
+			if err != nil {
+				return err
+			}
+			cacheDir := filepath.Join(config.GetAbsCachePath(), layout.ImagesDir)
+			location := filepath.Join(cacheDir, digest.Hex)
```

Comment:
The file path ends up looking like `.zarf-cache/images/sha256:779edd1984138d80000e2e7ca5deb8ee927712e1edc1426ec165bd156ff08c96` so we need the `sha256:` on top of the digest

Reply:
whoops, I used `.Hex` instead of `.String`, good catch